xla/service/collective_quantizer.cc

```diff
@@ -50,14 +60,121 @@ auto ConvertToNarrowerType() {
+        ShapeUtil::MakeShapeWithDenseLayout(
+            instr->shape().element_type(), unary_op->shape().dimensions(),
+            unary_op->shape().layout().minor_to_major()),
+        {instr}));
+  }
+  return instr;
+}
+
+// Recursively collects unary, divide, or multiply operands of instr until a
+// conversion to a wider type is reached. Returns an empty vector when no
+// conversion is reached.
+std::vector<HloInstruction*> FindDequantizationSubgraphRecursive(
+    HloInstruction* instr, absl::flat_hash_set<int>& visited_instrs,
+    std::vector<HloInstruction*> subgraph) {
+  // Avoid visiting the same instruction more than once.
+  if (!visited_instrs.emplace(instr->unique_id()).second) {
+    return {};
+  }
+
+  subgraph.emplace_back(instr);
+  if (Match(instr, ConvertToWiderType())) {
+    return subgraph;
+  }
+  if (instr->operand_count() == 1 || instr->opcode() == HloOpcode::kDivide) {
+    return FindDequantizationSubgraphRecursive(instr->mutable_operand(0),
+                                               visited_instrs, subgraph);
+  } else if (instr->opcode() == HloOpcode::kMultiply) {
+    for (HloInstruction* operand : instr->unique_operands()) {
+      auto binary_subgraph = FindDequantizationSubgraphRecursive(
+          operand, visited_instrs, subgraph);
+      if (!binary_subgraph.empty()) {
+        return binary_subgraph;
+      }
+    }
+  }
+  return {};
+}
+
+// Returns true iff instr describes a dequantization, i.e. a multiplication or
+// division by a broadcasted scalar operating on a type conversion, or a plain
+// type conversion to a wider type. Unary bitcast, copy, reshape or slice ops
+// may follow the dequantization or type conversion.
+std::optional<ConversionSubgraph> IsSupportedDequantization(
+    HloInstruction* instr) {
+  ConversionSubgraph subgraph;
+  absl::flat_hash_set<int> visited_instrs;
+  std::vector<HloInstruction*> candidate_subgraph =
+      FindDequantizationSubgraphRecursive(instr, visited_instrs,
+                                          std::vector<HloInstruction*>{});
+  std::reverse(candidate_subgraph.begin(), candidate_subgraph.end());
+
+  // In the dequantization case, the type conversion is followed by a
+  // multiplication or division by a broadcasted scalar.
+  if (candidate_subgraph.size() > 1 &&
+      (Match(
+           candidate_subgraph[1],
+           m::MultiplyAnyOrder(&subgraph.binary, m::Convert(&subgraph.convert),
+                               ScalarBroadcast(&subgraph.scale_bcast))) ||
+       Match(candidate_subgraph[1],
+             m::Divide(&subgraph.binary, m::Convert(&subgraph.convert),
+                       ScalarBroadcast(&subgraph.scale_bcast))))) {
+    subgraph.unaries = {candidate_subgraph.begin() + 2,
+                        candidate_subgraph.end()};
+  } else if (candidate_subgraph.size() > 0 &&
+             Match(candidate_subgraph[0], m::Convert(&subgraph.convert))) {
+    subgraph.unaries = {candidate_subgraph.begin() + 1,
```

Comment:
According to [this StackOverflow quesiton](https://stackoverflow.com/questions/60112841/copying-structs-with-uninitialized-members), copying a struct with uninitialized members is undefined behavior. So initialize all members of subgraph to avoid this. The best way is to just set them to nullptr in the definition of ConversionSubgraph.